File: infra/src/stacks/shared.ts
import * as gcp from "@pulumi/gcp"
import * as pulumi from "@pulumi/pulumi"
import { getSharedStackConfig } from "../config"

const config = getSharedStackConfig()

// Service account email is deterministic based on accountId and projectId
const githubActionsSaEmail = `github-actions-service@${config.projectId}.iam.gserviceaccount.com`

/**
 * APIs required for Cloud Run, Artifact Registry, Load Balancing, and IAM.
 */
const REQUIRED_APIS = [
  "run.googleapis.com",
  "compute.googleapis.com",
  "iam.googleapis.com",
  "cloudresourcemanager.googleapis.com",
  "artifactregistry.googleapis.com",
  "cloudbuild.googleapis.com",
  "iap.googleapis.com",
  "monitoring.googleapis.com",
  "logging.googleapis.com",
  "certificatemanager.googleapis.com",
]

// ========================================
// Enable Required APIs
// ========================================

const apiServices = REQUIRED_APIS.map(
  (api) =>
    new gcp.projects.Service(
      `api-${api.replace(/\./g, "-")}`,
      {
        project: config.projectId,
        service: api,
        disableDependentServices: false,
        disableOnDestroy: false,
      },
      {
        // Import existing API service from CDKTF state
        import: `${config.projectId}/${api}`,
      }
    )
)

// ========================================
// Artifact Registry
// ========================================

const artifactRegistry = new gcp.artifactregistry.Repository(
  "artifact-registry",
  {
    project: config.projectId,
    location: "asia-northeast1",
    repositoryId: "koborin-ai-web",
    description: "Container images for koborin.ai web application (dev/prod)",
    format: "DOCKER",
    dockerConfig: {
      immutableTags: true,
    },
  },
  {
    dependsOn: apiServices,
    // Import existing Artifact Registry from CDKTF state
    import: `projects/${config.projectId}/locations/asia-northeast1/repositories/koborin-ai-web`,
  }
)

// ========================================
// Global Static IP
// ========================================

const staticIp = new gcp.compute.GlobalAddress(
  "global-ip",
  {
    project: config.projectId,
    name: "koborin-ai-global-ip",
    addressType: "EXTERNAL",
    ipVersion: "IPV4",
    description: "Static IP for koborin.ai HTTPS load balancer",
  },
  {
    dependsOn: apiServices,
    // Import existing Global Address from CDKTF state
    import: `projects/${config.projectId}/global/addresses/koborin-ai-global-ip`,
  }
)

// ========================================
// Dev Environment Backend
// ========================================

// Serverless NEG for dev Cloud Run
const devNeg = new gcp.compute.RegionNetworkEndpointGroup(
  "dev-neg",
  {
    project: config.projectId,
    region: "asia-northeast1",
    name: "koborin-ai-dev-neg",
    networkEndpointType: "SERVERLESS",
    cloudRun: {
      service: "koborin-ai-web-dev",
    },
  },
  {
    dependsOn: apiServices,
    // Import existing NEG from CDKTF state
    import: `projects/${config.projectId}/regions/asia-northeast1/networkEndpointGroups/koborin-ai-dev-neg`,
  }
)

// Backend Service for dev (with IAP)
const devBackend = new gcp.compute.BackendService(
  "dev-backend",
  {
    project: config.projectId,
    name: "koborin-ai-dev-backend",
    protocol: "HTTP",
    loadBalancingScheme: "EXTERNAL_MANAGED",
    timeoutSec: 30,
    customResponseHeaders: ["X-Robots-Tag: noindex, nofollow"],
    backends: [
      {
        group: devNeg.id,
        balancingMode: "UTILIZATION",
        capacityScaler: 1.0,
      },
    ],
    iap: {
      enabled: true,
      oauth2ClientId: config.oauthClientId,
      oauth2ClientSecret: config.oauthClientSecret,
    },
  },
  {
    dependsOn: [devNeg],
    // Import existing Backend Service from CDKTF state
    import: `projects/${config.projectId}/global/backendServices/koborin-ai-dev-backend`,
  }
)

// IAP access control for dev
const _devIapAccess = new gcp.iap.WebBackendServiceIamBinding(
  "dev-iap-access",
  {
    project: config.projectId,
    webBackendService: devBackend.name,
    role: "roles/iap.httpsResourceAccessor",
    members: [`user:${config.iapUser}`],
  },
  {
    dependsOn: [devBackend],
    // Import existing IAP binding from CDKTF state
    // Format: projects/{project}/iap_web/compute/services/{web_backend_service}
    import: `projects/${config.projectId}/iap_web/compute/services/koborin-ai-dev-backend`,
  }
)

// ========================================
// Prod Environment Backend
// ========================================

// Serverless NEG for prod Cloud Run
const prodNeg = new gcp.compute.RegionNetworkEndpointGroup(
  "prod-neg",
  {
    project: config.projectId,
    region: "asia-northeast1",
    name: "koborin-ai-prod-neg",
    networkEndpointType: "SERVERLESS",
    cloudRun: {
      service: "koborin-ai-web-prod",
    },
  },
  {
    dependsOn: apiServices,
    // Import existing NEG from CDKTF state
    import: `projects/${config.projectId}/regions/asia-northeast1/networkEndpointGroups/koborin-ai-prod-neg`,
  }
)

// Backend Service for prod (no IAP)
const prodBackend = new gcp.compute.BackendService(
  "prod-backend",
  {
    project: config.projectId,
    name: "koborin-ai-prod-backend",
    protocol: "HTTP",
    loadBalancingScheme: "EXTERNAL_MANAGED",
    timeoutSec: 30,
    backends: [
      {
        group: prodNeg.id,
        balancingMode: "UTILIZATION",
        capacityScaler: 1.0,
      },
    ],
    logConfig: {
      enable: true,
      sampleRate: 1.0,
    },
  },
  {
    dependsOn: [prodNeg],
    // Import existing Backend Service from CDKTF state
    import: `projects/${config.projectId}/global/backendServices/koborin-ai-prod-backend`,
  }
)

// ========================================
// HTTPS Load Balancer
// ========================================

// Managed SSL Certificate (multi-domain)
const sslCert = new gcp.compute.ManagedSslCertificate(
  "managed-cert",
  {
    project: config.projectId,
    name: "koborin-ai-cert",
    managed: {
      domains: ["koborin.ai", "dev.koborin.ai"],
    },
  },
  {
    dependsOn: apiServices,
    // Import existing SSL certificate from CDKTF state
    import: `projects/${config.projectId}/global/sslCertificates/koborin-ai-cert`,
  }
)

// URL Map (host-based routing)
const urlMap = new gcp.compute.URLMap(
  "url-map",
  {
    project: config.projectId,
    name: "koborin-ai-url-map",
    description: "Routes traffic to dev/prod backends based on host header",
    defaultService: prodBackend.id,
    hostRules: [
      { hosts: ["koborin.ai"], pathMatcher: "prod-matcher" },
      { hosts: ["dev.koborin.ai"], pathMatcher: "dev-matcher" },
    ],
    pathMatchers: [
      { name: "prod-matcher", defaultService: prodBackend.id },
      { name: "dev-matcher", defaultService: devBackend.id },
    ],
  },
  {
    dependsOn: [devBackend, prodBackend],
    // Import existing URL Map from CDKTF state
    import: `projects/${config.projectId}/global/urlMaps/koborin-ai-url-map`,
  }
)

// Target HTTPS Proxy
const httpsProxy = new gcp.compute.TargetHttpsProxy(
  "https-proxy",
  {
    project: config.projectId,
    name: "koborin-ai-https-proxy",
    urlMap: urlMap.id,
    sslCertificates: [sslCert.id],
  },
  {
    dependsOn: [urlMap, sslCert],
    // Import existing HTTPS proxy from CDKTF state
    import: `projects/${config.projectId}/global/targetHttpsProxies/koborin-ai-https-proxy`,
  }
)

// Global Forwarding Rule
const forwardingRule = new gcp.compute.GlobalForwardingRule(
  "forwarding-rule",
  {
    project: config.projectId,
    name: "koborin-ai-forwarding-rule",
    target: httpsProxy.id,
    portRange: "443",
    ipProtocol: "TCP",
    loadBalancingScheme: "EXTERNAL_MANAGED",
    networkTier: "PREMIUM",
    ipAddress: staticIp.address,
  },
  {
    dependsOn: [httpsProxy, staticIp],
    // Import existing Forwarding Rule from CDKTF state
    import: `projects/${config.projectId}/global/forwardingRules/koborin-ai-forwarding-rule`,
  }
)

// ========================================
// Workload Identity (for GitHub Actions)
// ========================================

// Workload Identity Pool
const workloadIdentityPool = new gcp.iam.WorkloadIdentityPool(
  "github-actions-pool",
  {
    project: config.projectId,
    workloadIdentityPoolId: "github-actions-pool",
    displayName: "github-actions-pool",
    description: "Workload Identity Pool for GitHub Actions workflows",
  },
  {
    // Import existing Workload Identity Pool from CDKTF state
    import: `projects/${config.projectId}/locations/global/workloadIdentityPools/github-actions-pool`,
  }
)

// Workload Identity Provider (GitHub OIDC)
const workloadIdentityProvider = new gcp.iam.WorkloadIdentityPoolProvider(
  "github-provider",
  {
    project: config.projectId,
    workloadIdentityPoolId: workloadIdentityPool.workloadIdentityPoolId,
    workloadIdentityPoolProviderId: "actions-firebase-provider",
    displayName: "github-actions-provider",
    description: "GitHub Actions OIDC provider",
    // Only allow workflows from repositories owned by nozomi-koborinai
    attributeCondition: 'assertion.repository_owner == "nozomi-koborinai"',
    attributeMapping: {
      "google.subject": "assertion.repository",
      "attribute.repository_owner": "assertion.repository_owner",
    },
    oidc: {
      issuerUri: "https://token.actions.githubusercontent.com",
    },
  },
  {
    // Import existing Workload Identity Provider from CDKTF state
    import: `projects/${config.projectId}/locations/global/workloadIdentityPools/github-actions-pool/providers/actions-firebase-provider`,
  }
)

// Service Account for Terraform deployment
const githubActionsSa = new gcp.serviceaccount.Account(
  "github-actions-sa",
  {
    project: config.projectId,
    accountId: "github-actions-service",
    displayName: "github-actions-service",
    description: "Service account for GitHub Actions to deploy via Terraform",
  },
  {
    // Import existing Service Account from CDKTF state
    import: `projects/${config.projectId}/serviceAccounts/${githubActionsSaEmail}`,
  }
)

// Allow Workload Identity Pool to impersonate the service account
const _githubWifUser = new gcp.serviceaccount.IAMMember(
  "github-wif-user",
  {
    serviceAccountId: githubActionsSa.name,
    role: "roles/iam.workloadIdentityUser",
    member: pulumi.interpolate`principal://iam.googleapis.com/projects/${config.projectNumber}/locations/global/workloadIdentityPools/${workloadIdentityPool.workloadIdentityPoolId}/subject/nozomi-koborinai/koborin-ai`,
  },
  {
    // Import existing IAM binding from CDKTF state
    import: `projects/${config.projectId}/serviceAccounts/${githubActionsSaEmail} roles/iam.workloadIdentityUser principal://iam.googleapis.com/projects/${config.projectNumber}/locations/global/workloadIdentityPools/github-actions-pool/subject/nozomi-koborinai/koborin-ai`,
  }
)

// Grant necessary roles to the deployer service account
const DEPLOYER_ROLES = [
  "roles/artifactregistry.admin",
  "roles/cloudbuild.builds.builder",
  "roles/cloudbuild.builds.viewer",
  "roles/run.admin",
  "roles/compute.admin",
  "roles/iap.admin",
  "roles/logging.admin",
  "roles/logging.viewer",
  "roles/monitoring.admin",
  "roles/resourcemanager.projectIamAdmin",
  "roles/iam.serviceAccountUser",
  "roles/iam.serviceAccountAdmin",
  "roles/iam.workloadIdentityPoolAdmin",
  "roles/serviceusage.serviceUsageAdmin",
  "roles/storage.objectAdmin",
]

const _deployerRoleBindings = DEPLOYER_ROLES.map(
  (role) =>
    new gcp.projects.IAMMember(
      `deployer-sa-${role.replace(/\./g, "-").replace(/\//g, "-")}`,
      {
        project: config.projectId,
        role: role,
        member: pulumi.interpolate`serviceAccount:${githubActionsSa.email}`,
      },
      {
        // Import existing IAM binding from CDKTF state
        import: `${config.projectId} ${role} serviceAccount:${githubActionsSaEmail}`,
      }
    )
)

// ========================================
// Exports
// ========================================

export const artifactRegistryId = artifactRegistry.id
export const staticIpAddress = staticIp.address
export const devBackendId = devBackend.id
export const prodBackendId = prodBackend.id
export const urlMapId = urlMap.id
export const httpsProxyId = httpsProxy.id
export const forwardingRuleId = forwardingRule.id
export const workloadIdentityPoolId = workloadIdentityPool.workloadIdentityPoolId
export const workloadIdentityProviderId = workloadIdentityProvider.workloadIdentityPoolProviderId
export const githubActionsServiceAccountEmail = githubActionsSa.email

File: infra/stacks/shared.go
package stacks

import (
	"strings"

	"github.com/pulumi/pulumi-gcp/sdk/v8/go/gcp/artifactregistry"
	"github.com/pulumi/pulumi-gcp/sdk/v8/go/gcp/compute"
	"github.com/pulumi/pulumi-gcp/sdk/v8/go/gcp/iam"
	"github.com/pulumi/pulumi-gcp/sdk/v8/go/gcp/iap"
	"github.com/pulumi/pulumi-gcp/sdk/v8/go/gcp/projects"
	"github.com/pulumi/pulumi-gcp/sdk/v8/go/gcp/serviceaccount"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi/config"
)

// Shared creates resources for the shared stack.
func Shared(ctx *pulumi.Context) error {
	cfg := config.New(ctx, "")
	gcpCfg := config.New(ctx, "gcp")

	projectID := gcpCfg.Require("project")
	projectNumber := cfg.Require("projectNumber")
	iapUser := cfg.Require("iapUser")
	oauthClientID := cfg.Require("oauthClientId")
	oauthClientSecret := cfg.RequireSecret("oauthClientSecret")

	// ========================================
	// Enable Required APIs
	// ========================================

	requiredAPIs := []string{
		"run.googleapis.com",
		"compute.googleapis.com",
		"iam.googleapis.com",
		"cloudresourcemanager.googleapis.com",
		"artifactregistry.googleapis.com",
		"cloudbuild.googleapis.com",
		"iap.googleapis.com",
		"monitoring.googleapis.com",
		"logging.googleapis.com",
		"certificatemanager.googleapis.com",
	}

	var apiServices []pulumi.Resource
	for _, api := range requiredAPIs {
		// Logical name: api-<api-with-dots-replaced-by-dashes>
		// e.g., "api-run-googleapis-com"
		logicalName := "api-" + strings.ReplaceAll(api, ".", "-")
		svc, err := projects.NewService(ctx, logicalName, &projects.ServiceArgs{
			Project:                  pulumi.String(projectID),
			Service:                  pulumi.String(api),
			DisableDependentServices: pulumi.Bool(false),
			DisableOnDestroy:         pulumi.Bool(false),
		})
		if err != nil {
			return err
		}
		apiServices = append(apiServices, svc)
	}

	// ========================================
	// Artifact Registry
	// ========================================

	_, err := artifactregistry.NewRepository(ctx, "artifact-registry", &artifactregistry.RepositoryArgs{
		Project:      pulumi.String(projectID),
		Location:     pulumi.String("asia-northeast1"),
		RepositoryId: pulumi.String("koborin-ai-web"),
		Description:  pulumi.String("Container images for koborin.ai web application (dev/prod)"),
		Format:       pulumi.String("DOCKER"),
		DockerConfig: &artifactregistry.RepositoryDockerConfigArgs{
			ImmutableTags: pulumi.Bool(true),
		},
	}, pulumi.DependsOn(apiServices))
	if err != nil {
		return err
	}

	// ========================================
	// Global Static IP
	// ========================================

	staticIP, err := compute.NewGlobalAddress(ctx, "global-ip", &compute.GlobalAddressArgs{
		Project:     pulumi.String(projectID),
		Name:        pulumi.String("koborin-ai-global-ip"),
		AddressType: pulumi.String("EXTERNAL"),
		IpVersion:   pulumi.String("IPV4"),
		Description: pulumi.String("Static IP for koborin.ai HTTPS load balancer"),
	}, pulumi.DependsOn(apiServices))
	if err != nil {
		return err
	}

	// ========================================
	// Dev Environment Backend
	// ========================================

	// Serverless NEG for dev Cloud Run
	devNEG, err := compute.NewRegionNetworkEndpointGroup(ctx, "dev-neg", &compute.RegionNetworkEndpointGroupArgs{
		Project:             pulumi.String(projectID),
		Region:              pulumi.String("asia-northeast1"),
		Name:                pulumi.String("koborin-ai-dev-neg"),
		NetworkEndpointType: pulumi.String("SERVERLESS"),
		CloudRun: &compute.RegionNetworkEndpointGroupCloudRunArgs{
			Service: pulumi.String("koborin-ai-web-dev"),
		},
	}, pulumi.DependsOn(apiServices))
	if err != nil {
		return err
	}

	// Backend Service for dev (with IAP)
	devBackend, err := compute.NewBackendService(ctx, "dev-backend", &compute.BackendServiceArgs{
		Project:              pulumi.String(projectID),
		Name:                 pulumi.String("koborin-ai-dev-backend"),
		Protocol:             pulumi.String("HTTP"),
		LoadBalancingScheme:  pulumi.String("EXTERNAL_MANAGED"),
		TimeoutSec:           pulumi.Int(30),
		CustomResponseHeaders: pulumi.StringArray{pulumi.String("X-Robots-Tag: noindex, nofollow")},
		Backends: compute.BackendServiceBackendArray{
			&compute.BackendServiceBackendArgs{
				Group:          devNEG.ID(),
				BalancingMode:  pulumi.String("UTILIZATION"),
				CapacityScaler: pulumi.Float64(1.0),
			},
		},
		Iap: &compute.BackendServiceIapArgs{
			Enabled:            pulumi.Bool(true),
			Oauth2ClientId:     pulumi.String(oauthClientID),
			Oauth2ClientSecret: oauthClientSecret,
		},
	}, pulumi.DependsOn([]pulumi.Resource{devNEG}))
	if err != nil {
		return err
	}

	// IAP access control for dev
	_, err = iap.NewWebBackendServiceIamBinding(ctx, "dev-iap-access", &iap.WebBackendServiceIamBindingArgs{
		Project:           pulumi.String(projectID),
		WebBackendService: devBackend.Name,
		Role:              pulumi.String("roles/iap.httpsResourceAccessor"),
		Members:           pulumi.StringArray{pulumi.Sprintf("user:%s", iapUser)},
	}, pulumi.DependsOn([]pulumi.Resource{devBackend}))
	if err != nil {
		return err
	}

	// ========================================
	// Prod Environment Backend
	// ========================================

	// Serverless NEG for prod Cloud Run
	prodNEG, err := compute.NewRegionNetworkEndpointGroup(ctx, "prod-neg", &compute.RegionNetworkEndpointGroupArgs{
		Project:             pulumi.String(projectID),
		Region:              pulumi.String("asia-northeast1"),
		Name:                pulumi.String("koborin-ai-prod-neg"),
		NetworkEndpointType: pulumi.String("SERVERLESS"),
		CloudRun: &compute.RegionNetworkEndpointGroupCloudRunArgs{
			Service: pulumi.String("koborin-ai-web-prod"),
		},
	}, pulumi.DependsOn(apiServices))
	if err != nil {
		return err
	}

	// Backend Service for prod (no IAP)
	prodBackend, err := compute.NewBackendService(ctx, "prod-backend", &compute.BackendServiceArgs{
		Project:             pulumi.String(projectID),
		Name:                pulumi.String("koborin-ai-prod-backend"),
		Protocol:            pulumi.String("HTTP"),
		LoadBalancingScheme: pulumi.String("EXTERNAL_MANAGED"),
		TimeoutSec:          pulumi.Int(30),
		Backends: compute.BackendServiceBackendArray{
			&compute.BackendServiceBackendArgs{
				Group:          prodNEG.ID(),
				BalancingMode:  pulumi.String("UTILIZATION"),
				CapacityScaler: pulumi.Float64(1.0),
			},
		},
		LogConfig: &compute.BackendServiceLogConfigArgs{
			Enable:     pulumi.Bool(true),
			SampleRate: pulumi.Float64(1.0),
		},
	}, pulumi.DependsOn([]pulumi.Resource{prodNEG}))
	if err != nil {
		return err
	}

	// ========================================
	// HTTPS Load Balancer
	// ========================================

	// Managed SSL Certificate (multi-domain)
	sslCert, err := compute.NewManagedSslCertificate(ctx, "managed-cert", &compute.ManagedSslCertificateArgs{
		Project: pulumi.String(projectID),
		Name:    pulumi.String("koborin-ai-cert"),
		Managed: &compute.ManagedSslCertificateManagedArgs{
			Domains: pulumi.StringArray{
				pulumi.String("koborin.ai"),
				pulumi.String("dev.koborin.ai"),
			},
		},
	}, pulumi.DependsOn(apiServices))
	if err != nil {
		return err
	}

	// URL Map (host-based routing)
	urlMap, err := compute.NewURLMap(ctx, "url-map", &compute.URLMapArgs{
		Project:        pulumi.String(projectID),
		Name:           pulumi.String("koborin-ai-url-map"),
		Description:    pulumi.String("Routes traffic to dev/prod backends based on host header"),
		DefaultService: prodBackend.ID(),
		HostRules: compute.URLMapHostRuleArray{
			&compute.URLMapHostRuleArgs{
				Hosts:       pulumi.StringArray{pulumi.String("koborin.ai")},
				PathMatcher: pulumi.String("prod-matcher"),
			},
			&compute.URLMapHostRuleArgs{
				Hosts:       pulumi.StringArray{pulumi.String("dev.koborin.ai")},
				PathMatcher: pulumi.String("dev-matcher"),
			},
		},
		PathMatchers: compute.URLMapPathMatcherArray{
			&compute.URLMapPathMatcherArgs{
				Name:           pulumi.String("prod-matcher"),
				DefaultService: prodBackend.ID(),
			},
			&compute.URLMapPathMatcherArgs{
				Name:           pulumi.String("dev-matcher"),
				DefaultService: devBackend.ID(),
			},
		},
	}, pulumi.DependsOn([]pulumi.Resource{devBackend, prodBackend}))
	if err != nil {
		return err
	}

	// Target HTTPS Proxy
	httpsProxy, err := compute.NewTargetHttpsProxy(ctx, "https-proxy", &compute.TargetHttpsProxyArgs{
		Project: pulumi.String(projectID),
		Name:    pulumi.String("koborin-ai-https-proxy"),
		UrlMap:  urlMap.ID(),
		SslCertificates: pulumi.StringArray{
			sslCert.ID(),
		},
	}, pulumi.DependsOn([]pulumi.Resource{urlMap, sslCert}))
	if err != nil {
		return err
	}

	// Global Forwarding Rule
	_, err = compute.NewGlobalForwardingRule(ctx, "forwarding-rule", &compute.GlobalForwardingRuleArgs{
		Project:             pulumi.String(projectID),
		Name:                pulumi.String("koborin-ai-forwarding-rule"),
		Target:              httpsProxy.ID(),
		PortRange:           pulumi.String("443"),
		IpProtocol:          pulumi.String("TCP"),
		LoadBalancingScheme: pulumi.String("EXTERNAL_MANAGED"),
		NetworkTier:         pulumi.String("PREMIUM"),
		IpAddress:           staticIP.Address,
	}, pulumi.DependsOn([]pulumi.Resource{httpsProxy, staticIP}))
	if err != nil {
		return err
	}

	// ========================================
	// Workload Identity (for GitHub Actions)
	// ========================================

	// Workload Identity Pool
	workloadIdentityPool, err := iam.NewWorkloadIdentityPool(ctx, "github-actions-pool", &iam.WorkloadIdentityPoolArgs{
		Project:                pulumi.String(projectID),
		WorkloadIdentityPoolId: pulumi.String("github-actions-pool"),
		DisplayName:            pulumi.String("github-actions-pool"),
		Description:            pulumi.String("Workload Identity Pool for GitHub Actions workflows"),
	})
	if err != nil {
		return err
	}

	// Workload Identity Provider (GitHub OIDC)
	_, err = iam.NewWorkloadIdentityPoolProvider(ctx, "github-provider", &iam.WorkloadIdentityPoolProviderArgs{
		Project:                        pulumi.String(projectID),
		WorkloadIdentityPoolId:         workloadIdentityPool.WorkloadIdentityPoolId,
		WorkloadIdentityPoolProviderId: pulumi.String("actions-firebase-provider"),
		DisplayName:                    pulumi.String("github-actions-provider"),
		Description:                    pulumi.String("GitHub Actions OIDC provider"),
		// Only allow workflows from repositories owned by nozomi-koborinai
		AttributeCondition: pulumi.String(`assertion.repository_owner == "nozomi-koborinai"`),
		AttributeMapping: pulumi.StringMap{
			"google.subject":             pulumi.String("assertion.repository"),
			"attribute.repository_owner": pulumi.String("assertion.repository_owner"),
		},
		Oidc: &iam.WorkloadIdentityPoolProviderOidcArgs{
			IssuerUri: pulumi.String("https://token.actions.githubusercontent.com"),
		},
	})
	if err != nil {
		return err
	}

	// Service Account for Pulumi deployment
	githubActionsSA, err := serviceaccount.NewAccount(ctx, "github-actions-sa", &serviceaccount.AccountArgs{
		Project:     pulumi.String(projectID),
		AccountId:   pulumi.String("github-actions-service"),
		DisplayName: pulumi.String("github-actions-service"),
		Description: pulumi.String("Service account for GitHub Actions to deploy via Pulumi"),
	})
	if err != nil {
		return err
	}

	// Allow Workload Identity Pool to impersonate the service account
	_, err = serviceaccount.NewIAMMember(ctx, "github-wif-user", &serviceaccount.IAMMemberArgs{
		ServiceAccountId: githubActionsSA.Name,
		Role:             pulumi.String("roles/iam.workloadIdentityUser"),
		Member: pulumi.Sprintf(
			"principal://iam.googleapis.com/projects/%s/locations/global/workloadIdentityPools/%s/subject/nozomi-koborinai/koborin-ai",
			projectNumber,
			workloadIdentityPool.WorkloadIdentityPoolId,
		),
	})
	if err != nil {
		return err
	}

	// Grant necessary roles to the deployer service account
	deployerRoles := []string{
		"roles/artifactregistry.admin",
		"roles/cloudbuild.builds.builder",
		"roles/cloudbuild.builds.viewer",
		"roles/run.admin",
		"roles/compute.admin",
		"roles/iap.admin",
		"roles/logging.admin",
		"roles/logging.viewer",
		"roles/monitoring.admin",
		"roles/resourcemanager.projectIamAdmin",
		"roles/iam.serviceAccountUser",
		"roles/iam.serviceAccountAdmin",
		"roles/iam.workloadIdentityPoolAdmin",
		"roles/serviceusage.serviceUsageAdmin",
		"roles/storage.objectAdmin",
	}

	for _, role := range deployerRoles {
		// Logical name: deployer-sa-<role-with-dots-and-slashes-replaced-by-dashes>
		// e.g., "deployer-sa-roles-artifactregistry-admin"
		logicalName := "deployer-sa-" + strings.ReplaceAll(strings.ReplaceAll(role, ".", "-"), "/", "-")
		_, err = projects.NewIAMMember(ctx, logicalName, &projects.IAMMemberArgs{
			Project: pulumi.String(projectID),
			Role:    pulumi.String(role),
			Member:  pulumi.Sprintf("serviceAccount:%s", githubActionsSA.Email),
		})
		if err != nil {
			return err
		}
	}

	return nil
}

